gpt4all-backend/llmodel.cpp

```diff
@@ -75,12 +77,20 @@ bool LLModel::Implementation::isImplementation(const Dlhandle &dl) {
     return dl.get<bool(uint32_t)>("is_g4a_backend_model_implementation");
 }
 
-const std::vector<LLModel::Implementation> &LLModel::Implementation::implementationList() {
+static std::vector<std::unique_ptr<LLModel::Implementation>> s_impl_libs;
+static bool s_scanned = false;
+
+const std::vector<std::unique_ptr<LLModel::Implementation>> &LLModel::Implementation::implementationList() {
     // NOTE: allocated on heap so we leak intentionally on exit so we have a chance to clean up the
     // individual models without the cleanup of the static list interfering
-    static auto* libs = new std::vector<Implementation>([] () {
-        std::vector<Implementation> fres;
-
+    if(!s_scanned) {
```

Comment:
s_impl_libs should be a function-local static.

If we're at all concerned about thread safety, we should use a function to compute s_impl_libs and initialize it on the same line as the declaration - C++ guarantees that the initializer will be run only once. Otherwise, we should at least put `if (s_scanned) { return s_impl_libs; }` at the top to reduce the indentation level.

Reply:
moving the decl/init into the fn, but the whole thing here is that we can't actually *compute* it (do the dlopen, etc.) at static init time because the search path may not actually be set yet (in the GUI case at least it is set by a _different_ static initializer in a different file and it _sometimes_ worked if that one ran first, but it is not guaranteed that it will)

if we want to _actually_ be threadsafe around this we can't have a global mutable search path

Reply:
Static local variables are not initialized until the first time control passes through their declaration. See [cppreference](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables).

Reply:
Let's chat about this. From what I can tell, you're both right. It should be static local, but the problem remains if that function is called before the global search path is set?

Reply:
Yeahh either way if the method to set the search path is used it needs to have the re-checking behavior if it's not clearly documented to be useless unless carried very early

I am not really clear on how this had been being called before that but I definitely caught this code running and loading libs *before* the chat gui had changed the pathb from the default `.`

Reply:
Why does the chat client call setImplementationsSearchPath in the constructor of LLM instead of main.cpp?

Reply:
hmm, yeah, all the globalInstance()s should also be deferred til first use so *should* get run in the order they are called, and at least spot-checking it it does set the search path before its read when running from the build tree - built an installer to check and that _also_ remains the case - so its likely the bigger culprit here was it `dlopen`ing dlls it shouldn't and that check probably doing more work here than anything else